CI: Add upgrade test for PATCH VERSIONS

Ebaneck opened the pull request:
**Component**:

'ci', 'tests'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

See #2009 

**Summary**:

We should be able to automatically test patch versions using existing promoted artefacts.
Targeting 2.4 branches only because patch version inclusion was first introduced here.

**Acceptance criteria**: 

Please don't come with any more flakies

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2009

Ebaneck committed:
CI: Add upgrade testing for PATCH VERSIONS

This commit adds Upgrade testing for MEtalK8s PATCH versions.
Initially, we did not have any upgrade testing when a patch version
is introduced

closes: #2009

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` with contents from `feature/2009-add-steps-to-upgrade-between-patch-versions`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions origin/development/2.5
 $ git merge origin/feature/2009-add-steps-to-upgrade-between-patch-versions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions
```

alexandre-allard commented on eve/main.yml:
```diff
@@ -405,6 +405,14 @@ stages:
+              major=$(echo "%(prop:product_version)s" | cut -d'.' -f1) &&
+              minor=$(echo "%(prop:product_version)s" | cut -d'.' -f2) &&
+              patch=$(echo "%(prop:product_version)s" | cut -d'.' -f3) &&
+              echo "$major.$minor.$(( $patch-1 ))"
```
You must do some checks before doing that, otherwise you could end up with a version `2.5.-1`, not sure that what we want. :)
I would do something like:
```
              major=$(echo "%(prop:product_version)s" | cut -d'.' -f1) &&
              minor=$(echo "%(prop:product_version)s" | cut -d'.' -f2) &&
              patch=$(echo "%(prop:product_version)s" | cut -d'.' -f3) &&
              echo "$major.$minor.$(( patch > 0 ? patch - 1 : 0 ))"
```
And... I'm not sure how to handle it (except by commenting the whole `single-node-upgrade-patch-centos` step in 2.5.0), but we must not do patch version upgrade/downgrade tests if `product_patch_version_prev == product_version`.
Would be very nice to have the `doStepIf` for such cases.
Maybe we can leave it as is right now and comment the whole thing in 2.5.0, until we've a released patch version or a better way to activate/deactivate checks depending on variables.

Ebaneck replied:
As agreed in person, we will deactivate upgrade testing for patch versions on the 2.5 branch since `doStepIf` is buggy at this time.

gdemonet replied:
> Would be very nice to have the `doStepIf` for such cases.

Working on it :)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #2db0f303073da69bc1d0b67c412821208683f68d on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` with contents from `feature/2009-add-steps-to-upgrade-between-patch-versions`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions origin/development/2.5
 $ git merge origin/feature/2009-add-steps-to-upgrade-between-patch-versions
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #0edc00229e5a0ef5c26052a91d50cc9f234739e3 on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck committed:
salt: Hack to remove old nginx ingress control plane DS if needed

Make sure to remove immutable selectors for nginx ingress control
plane daemon set which were omitted.

Reference commit: 2ef50e492fc8e3ce9159e586bba57b13eff62fe0

alexandre-allard commented on eve/main.yml:
```diff
@@ -927,6 +928,129 @@ stages:
+          <<: *run_bootstrap
+          env: *_env_prev_patch_version
+      - ShellCommand:
+          name: Provision Prometheus and AlertManager storage
```
Not sure this step is really mandatory since we don't test anything after patch version installation, and you already create the volumes after the upgrade, before doing the tests.

Ebaneck replied:
Yes, it is not mandatory.
But if we finish the installation and try to upgrade immediately, we endup with cluster resources which are not ready.
BTW, during downgrade for patch versions, we will re-use this step

alexandre-allard commented on eve/main.yml:
```diff
@@ -927,6 +928,129 @@ stages:
+                      promoted-%(prop:product_version_prev_patch)s"
+      - ShellCommand: &retrieve_prev_patch_version_iso
+          <<: *retrieve_iso
+          name: Retrieve previous PATCH ISO image
```
detail: be consistent, use either `PATCH`, `Patch` or `patch` everywhere.

alexandre-allard reviewed (dismissed):
details, otherwise LGTM.

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2009-add-steps-to-upgrade-between-patch-versions` into
`development/2.4`
* [w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #b8d52f531fbfaff185c517d9372502e7a722f640 on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2009-add-steps-to-upgrade-between-patch-versions` into
`development/2.4`
* [w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #70b5809c5720f98ce1b5efb9a0fead18da8770e1 on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
/approve

bert-e commented:
# History mismatch



Merge commit #70b5809c5720f98ce1b5efb9a0fead18da8770e1 on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

alexandre-allard commented on eve/main.yml:
```diff
@@ -927,6 +928,131 @@ stages:
+              if [ "$PRODUCT_VERSION" != "$PRODUCT_VERSION_PREV_patch" ]; then
+                 echo "true"
+              else
+                echo "false"
```
detail: indentation is not consistent here

Ebaneck replied:
merde!

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2009-add-steps-to-upgrade-between-patch-versions` into
`development/2.4`
* [w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# History mismatch



Merge commit #c884c658df8a37e7efb254c6c4a240de915394eb on the integration branch
`w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions` is merging a branch which is neither the current
branch `feature/2009-add-steps-to-upgrade-between-patch-versions` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/2009-add-steps-to-upgrade-between-patch-versions` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/2009-add-steps-to-upgrade-between-patch-versions` into
`development/2.4`
* [w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/2009-add-steps-to-upgrade-between-patch-versions) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
